### PR TITLE
KMQ Premium

### PR DESCRIPTION
**Pull request by taahamahdi**
* More songs
* Special in any server
* Supporter badge

### CODE REVIEW

File: src/events/client/voiceChannelLeave.ts [hunk 1]
```diff
@@ -22,7 +27,8 @@ export default function voiceChannelLeaveHandler(
         return;
     }
 
-    if (!gameSession.finished) {
```
Comment (taahamahdi) [inline src/events/client/voiceChannelLeave.ts:25]: Add back this check

File: src/events/client/voiceChannelSwitch.ts [hunk 2]
```diff
@@ -24,7 +25,18 @@ export default async function voiceChannelSwitchHandler(
         return;
     }
 
-    if (!gameSession.finished) {
```
Comment (taahamahdi) [inline src/events/client/voiceChannelSwitch.ts:27]: Add back

File: src/helpers/patreon_manager.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import { Campaign } from "patreon-discord";
+import dbContext from "../database_context";
+import { IPCLogger } from "../logger";
+import { addPremium, removePremium } from "./game_utils";
+
+const logger = new IPCLogger("patreon_manager");
+
+export const PATREON_SUPPORTER_BADGE = "🎧 Premium Supporter";
+
+const campaign = new Campaign({
```
Comment (taahamahdi) [inline src/helpers/patreon_manager.ts:10]: Tie to state

File: src/helpers/patreon_manager.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+/**
+ * Fetch up-to-date Patreon members and update Premium members accordingly
+ */
+export default async function updatePremiumUsers(): Promise<void> {
+    let fetchedPatrons: Array<PatronResponse>;
+    try {
+        fetchedPatrons = await campaign.fetchPatrons([
+            "active_patron",
```
Comment (taahamahdi) [inline src/helpers/patreon_manager.ts:34]: Enum

File: src/structures/game_session.ts [hunk 5]
```diff
@@ -1181,6 +1193,57 @@ export default class GameSession {
...
+            this.premiumGame = true;
+            return;
+        }
+
+        const voiceMembers = getCurrentVoiceMembers(this.voiceChannelID);
+        for (const member of voiceMembers) {
+            if (await isUserPremium(member.id)) {
+                if (!this.premiumGame) {
+                    this.premiumGame = true;
+                    const guildPreference = await getGuildPreference(
```
Comment (taahamahdi) [inline src/structures/game_session.ts:1211]: Remove as variable

File: src/structures/game_session.ts [hunk 6]
```diff
@@ -1181,6 +1193,55 @@ export default class GameSession {
         this.addBookmarkedSong(interaction.member?.id, song);
     }
 
+    /**
+     * If the game changes its premium state, update filtered songs
+     * @param premiumJoined - true if a premium member joined VC
+     */
+    async updatePremiumStatus(premiumJoined: boolean): Promise<void> {
+        if (premiumJoined) {
+            this.premiumGame = true;
```
Comment (Brainicism) [inline src/structures/game_session.ts:1202]: Don't we still need to reload songs if game wasn't premium before this?

File: src/kmq_worker.ts [hunk 7]
```diff
@@ -117,6 +119,12 @@ export class BotWorker extends BaseClusterWorker {
             logger.info("Initializing bot stats poster...");
             const botListingManager = new BotListingManager();
             botListingManager.start();
+
+            logger.info("Initializing Patreon manager...");
+            state.patreonCampaign = new Campaign({
```
Comment (Brainicism) [inline src/kmq_worker.ts:124]: need nullcheck on these

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/structures/game_session.ts (hunk 6 above)
Comment (taahamahdi) [reply src/structures/game_session.ts]: Somewhat improved this function

File: docs/GAMEPLAY.md [hunk 8]
```diff
@@ -211,6 +211,12 @@ To keep things fair, switching teams mid-game forfeits your current points and E
 
 ![teams](https://raw.githubusercontent.com/Brainicism/KMQ_Discord/master/images/teams.png)
 
+# KMQ Premium
```
Comment (Brainicism) [inline docs/GAMEPLAY.md:214]: Let's not put anything on the docs until we release it officially

File: src/events/client/voiceChannelSwitch.ts [hunk 9]
```diff
@@ -31,18 +31,21 @@ export default async function voiceChannelSwitchHandler(
...
-            await gameSession.setPlayerInVC(
-                member.id,
-                newChannel.id === gameSession.voiceChannelID
-            );
-        } else {
-            // Bot was moved to another VC
-            gameSession.voiceChannelID = newChannel.id;
-            gameSession.syncAllVoiceMembers();
-        }
+    const oldPremiumState = gameSession.isPremiumGame();
```
Comment (Brainicism) [inline src/events/client/voiceChannelSwitch.ts:34]: Why isn't this on line 45?

File: src/helpers/game_utils.ts [hunk 10]
```diff
@@ -42,14 +43,16 @@ export async function ensureVoiceConnection(
 
 /**
  * @param guildPreference - The GuildPreference
+ * @param premium - Whether to include premium songs
  * @returns an object containing the total number of available songs before and after limit based on the GameOptions
  */
 export async function getAvailableSongCount(
-    guildPreference: GuildPreference
+    guildPreference: GuildPreference,
+    premium: boolean
```
Comment (Brainicism) [inline src/helpers/game_utils.ts:51]: isPremium

File: src/helpers/game_utils.ts [hunk 11]
```diff
@@ -410,6 +413,85 @@ export async function getMultipleChoiceOptions(
...
+ * @param guildID - The guild ID
+ * @param playerID - The player ID
+ * @returns whether the current game is a premium game, or the player is premium
+ */
+export async function isPremiumRequest(
+    guildID: string,
+    playerID: string
+): Promise<boolean> {
+    return (
+        state.gameSessions[guildID]?.isPremiumGame() ??
```
Comment (Brainicism) [inline src/helpers/game_utils.ts:490]: Will this cause confusion for active players if a premium player uses `,options` while he's not in an active game?

File: src/helpers/patreon_manager.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        .filter((x: PatronResponse) => !!x.discord_user_id)
+        .map((x: PatronResponse) => ({
+            discordID: x.discord_user_id,
+            activePatron: x.patron_status === PatronState.ACTIVE,
+            firstSubscribed: x.pledge_relationship_start,
+        }));
+
+    const activePatronIDs: string[] = patrons
+        .filter((x) => x.activePatron)
+        .map((x) => x.discordID);
```
Comment (Brainicism) [inline src/helpers/patreon_manager.ts:60]: Didn't you already filter out null discord ids?

File: src/helpers/patreon_manager.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+    removePremium(
```
Comment (Brainicism) [inline src/helpers/patreon_manager.ts:62]: We should still support premium players who weren't patrons. Need to find away to not delete those instances

File: src/migrations/20210821155505_premium_users.js [hunk 14]
```diff
@@ -0,0 +1,12 @@
+exports.up = function (knex) {
+    return knex.schema.createTable("premium_users", (table) => {
+        table.string("user_id");
+        table.boolean("active").notNullable();
+        table.datetime("first_subscribed");
+        table.primary(["user_id"]);
```
Comment (Brainicism) [inline src/migrations/20210821155505_premium_users.js:6]: Add another column to indicate how they got premium? patreon vs loyalty

File: src/structures/song_selector.ts [hunk 15]
```diff
@@ -215,9 +215,13 @@ export default class SongSelector {
         return this.filteredSongs.songs.size;
     }
 
-    async reloadSongs(guildPreference: GuildPreference): Promise<void> {
+    async reloadSongs(
+        guildPreference: GuildPreference,
+        premium: boolean
```
Comment (Brainicism) [inline src/structures/song_selector.ts:220]: isPremium

File: src/test/ci/game_utils.test.ts [hunk 16]
```diff
@@ -279,7 +279,11 @@ describe("song query", () => {
         it("should return the expected song count", async () => {
             const limit = 50;
             await guildPreference.setLimit(0, limit);
-            const songCount = await getAvailableSongCount(guildPreference);
+            const songCount = await getAvailableSongCount(
```
Comment (Brainicism) [inline src/test/ci/game_utils.test.ts:282]: Add test to compare premium vs non premium

File: src/test/ci/song_selector.test.ts [hunk 17]
```diff
@@ -85,7 +87,8 @@ describe("getFilteredSongList", () => {
                 const selectedArtist = { id: 208, name: "Twice" };
                 await guildPreference.setGroups([selectedArtist]);
                 const { songs } = await SongSelector.getFilteredSongList(
-                    guildPreference
+                    guildPreference,
+                    true
```
Comment (Brainicism) [inline src/test/ci/song_selector.test.ts:91]: Can we add a default value?

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: src/helpers/game_utils.ts (hunk 11 above)
Comment (taahamahdi) [reply src/helpers/game_utils.ts:490]: Don't think so. The options message has a gold embed color when it's a premium request, so it should be easily differentiated

File: src/helpers/patreon_manager.ts (hunk 12 above)
Comment (taahamahdi) [reply src/helpers/patreon_manager.ts:60]: This isn't filtering out non-null discord ID's, just getting the discord ID's from the patrons object

File: src/helpers/game_utils.ts (hunk 11 above)
Comment (Brainicism) [reply src/helpers/game_utils.ts:490]: Won't the people in the active game think their game is premium?

File: src/test/ci/song_selector.test.ts (hunk 17 above)
Comment (taahamahdi) [reply src/test/ci/song_selector.test.ts:91]: Should the default be premium = true or premium = false?

File: src/helpers/game_utils.ts (hunk 11 above)
Comment (taahamahdi) [reply src/helpers/game_utils.ts:490]: Potentially, yes. Should we use a separate color to show it's a premium game, and use that in end round messages? Maybe blue?

File: src/test/ci/song_selector.test.ts (hunk 17 above)
Comment (Brainicism) [reply src/test/ci/song_selector.test.ts:91]: false

**Review by Brainicism — APPROVED** [review]

**Pull request merged**